ARM: cortex-M: qemu: add support for executing tests with FPU/FPU_SHARING

ioannisg opened the pull request:
We are adding support for running tests and samples in CI for QEMU on Cortex-M with FPU and FPU_SHARING options enabled. **This fixes a severe limitation we had until now, by not covering at all the ARCH/Kernel implementation around Floating Point support in QEMU, at least not for Cortex-M. This is important for Zephyr LTS release.**

The solution is a bit complex, and involves extending the remote-core (CPU1) target of MPS2 AN521, since FPU support is only emulated on CPU1. The patch re-organizes a bit the AN521 SoC and board(s) definitions accordingly.

Effectively, we add support (locally for AN521) for multi-image builds, because the CPU1 binaries cannot run as standalone images. While it is desired to have a generic multi-image solution, we currently restricted the implementation to the AN521 board, to address our (important) use-case, without making drastic changes to Zephyr build framework.

ioannisg committed:
arm: cortex-m: add FPU option for v8.M without DSP

With respect to the mapping between the CPU variant
and the FPU we add an entry for Cortex-M v8-M CPU
variants without DSP. This should cover the case
of a Mainline Cortex-M which implements the Floating
Point extension but not the DSP extension.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
soc: arm: mps2_an521: have separate configs for the 2 cores of AN521

MPS2 AN521 is a dual (Cortex-M33) core where the two cores
have different capabilities. E.g CPU0 implements TrustZone-M,
while CPU1 implements the FPU. Therefore we introcude separate
Kconfig symbols for the two SoC variants, and use these to
select the different Cortex-M capabilities. We also update
the definitions of __MPU/FPU/SAU/DSP variant in soc.h to make
them get the value from the Kconfig settings, directly.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: select SOC_MPS2_AN521_CPU0 in mps2 targets on CPU0

Select Kconfig option SOC_MPS2_AN521_CPU0 in board
definitions of MPS2 AN521 targets which build on
CPU0 of the MPS2 AN521. This is the case for the
mps2_an521 and mps2_an521_nonsecure targets.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: add option to build a CPU0 image for CPU1 applications

We add an option to build a CPU0 image, when we are building
applications for AN521 CPU1 core. This image will be an empty
binary that will basically boot the device and wake up CPU1.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: separate board Kconfig options for the various targets

Introduce separate Kconfig options for the different
MPS2 AN521 targets.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: no empty CPU0 binary for openamp applications

When we are building applications with openAMP on CPU1, we do not
need to trigger a build an empty binary for CPU0, as this will be
take care of by the dual core sample anyways.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: add arm, and fpu tag for mps2_an521_remote

Enable testing Arm architecture and Kernel unit tests on
the mps2_an521_remote platform, with CONFIG_FPU and
CONFIG_FPU_SHARING enabled. This is done by enabling the
arm and fpu tags on the mps2_an521_nonsecure board target.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,24 @@
+
+  set(CPU0_BINARY_DIR ${BOARD_DIR}/empty_cpu0-prefix/src/empty-cpu0-build/zephyr)
+
+  enable_language(C ASM)
```
why is this needed here ?

ioannisg replied:
possibly some redundant copy-paste
will check.

ioannisg replied:
Removed now, thanks!

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,24 @@
+
+  ExternalProject_Add(
+    empty_cpu0
+    SOURCE_DIR ${BOARD_DIR}/empty_cpu0
```
I don't find `empty_cpu0` anywhere ?

ioannisg replied:
thanks! fixed

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,24 @@
+
+if(CONFIG_SOC_MPS2_AN521_CPU1)
+# Building a firmware image for CPU1: this requires a binary
+# for CPU0, which will boot the device and wake up CPU1.
```
If user builds a dedicated sample for CPU0 himself, then how can he avoid building the `empty_cpu0` sample ?

ioannisg replied:
Looks, like we cannot get around without a new Kconfig option indicating dual core builds, e.g. config BOARD_MPS2_AN521_BUILD_FOR_STANDALONE_CPU1

tejlmand commented on boards/arm/mps2_an521/board.cmake:
```diff
@@ -17,4 +17,9 @@ if (CONFIG_BUILD_WITH_TFM)
   set(QEMU_KERNEL_OPTION "-device;loader,file=${CMAKE_BINARY_DIR}/tfm_merged.hex")
+
+elseif (CONFIG_SOC_MPS2_AN521_CPU1)
+  set(CPU0_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/zephyr/boards/arm/mps2_an521/empty_cpu0-prefix/src/empty_cpu0-build/zephyr)
```
This looks something that could easily break in future.

ioannisg replied:
I leave this up to you, @tejlmand , to propose a better solution to this :)
For now I used what was already the solution in the rpmsg samples in the tree. 
Would welcome a proposal to improve this.

ioannisg replied:
@tejlmand any change you can provide a hint here?

tejlmand commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -8,9 +8,11 @@ zephyr_library()
 zephyr_library_include_directories(${ZEPHYR_BASE}/drivers)
 
-if(CONFIG_SOC_MPS2_AN521_CPU1)
+if(CONFIG_SOC_MPS2_AN521_CPU1 AND NOT CONFIG_OPENAMP)
```
should this be more generic ?
Afterall, OpenAmp might not be the only sample that should disable the empty-cpu0.

ioannisg replied:
Well, it could. Although, only openAMP samples/applications/tests for Cortex-M are currently multi-core.
Honestly, I would not introduce a new abstract option, since we only use this, currently, in the MPS2 AN521 platform.

ioannisg replied:
@tejlmand, I thought about it and my feeling is that we are OK with the comment that is placed here, inline.
We could have a CONFIG_DUAL_CORE_APPLICATION option to instruct the build, here, _not_ to build an empty CPU0, but since openamp is the only in-tree subsystem that currently relates to multi-core applications, I would skip introducing it in favor of having the explanatory comment.

tejlmand reviewed (commented):
The initial approach looks sane.

A couple of small observations.

ioannisg committed:
boards: mps2_an521: add empty CPU0 binary for AN521

Add a project to build an empty binary for AN521 CPU0,
which will only serve to wake up CPU1, and let samples,
and tests to execute as standalone applications in CPU1.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

stephanosio reviewed (approved):
It looks ok overall.

Just some notes:
1. cmsis-dsp tests should be updated to test with `mps2_an521_remote` (w/ hardfp) after this is merged (it is currently being tested using `mps2_an521` with softfp).
2. `v2m_musca_b1` has FPU enabled on the CPU0, so we could consider using it as the default v8M testing platform in the future:
https://github.com/qemu/qemu/blob/a97978bcc2d1f650c7d411428806e5b03082b8c7/hw/arm/musca.c#L390-L397

ioannisg committed:
tests: add fpu tag the tests which enable FPU and FPU_SHARING options

Introduce the fpu tag to tests that explicitly enable
the FPU and FPU_SHARING Kconfig options. The tag could
be used to run all FPU-related tests in the tree.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
> It looks ok overall.
> 
> Just some notes:
> 
> 1. cmsis-dsp tests should be updated to test with `mps2_an521_remote` (w/ hardfp) after this is merged (it is currently being tested using `mps2_an521` with softfp).
> 2. `v2m_musca_b1` has FPU enabled on the CPU0, so we could consider using it as the default v8M testing platform in the future:
>    https://github.com/qemu/qemu/blob/a97978bcc2d1f650c7d411428806e5b03082b8c7/hw/arm/musca.c#L390-L397

Opened ticket: https://github.com/zephyrproject-rtos/zephyr/issues/36013

ioannisg committed:
doc: release notes entry for FPU support in QEMU

A release note for adding FPU support in QEMU for
ARM Cortex-M.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: mps2_an521: update build instructions for CPU0 and CPU1

A major update to the MPS2 AN521 build instructions, covering
building with TF-M for CPU0 and building for CPU1.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

microbuilder commented on boards/arm/mps2_an521/Kconfig.board:
```diff
@@ -1,8 +1,20 @@
+	select QEMU_TARGET
+	select HAS_COVERAGE_SUPPORT
+
+config BOARD_MPS2_AN521_CPU0NS
```
also a nit, but this feels a bit unnatural to me, and `BOARD_MPS2_AN521_CPU0_NS` feels easier to read, and is more consistent with current `_ns` or `_nonsecure` naming conventions?

ioannisg replied:
the Nordic boards have `ns` and `NS` without the underscores, so we lack consistency. I 'll change this, anyways, to what you suggest. But i think we need to agree on some rules here for the Non-Secure variant naming.

ioannisg replied:
Fixed.

microbuilder commented on boards/arm/mps2_an521/doc/index.rst:
```diff
@@ -32,14 +32,36 @@ More information about the board can be found at the `MPS2 FPGA Website`_.
++----------------------+-------------------------------------------------------+
+| mps2_an521_nonsecure | For building Non-Secure firmware for CPU0             |
++----------------------+-------------------------------------------------------+
+| mps2_an521_remote    | For buildling firmware on CPU1                        |
```
```suggestion
| mps2_an521_remote    | For building firmware on CPU1                        |
```

microbuilder commented on boards/arm/mps2_an521/doc/index.rst:
```diff
@@ -320,12 +342,77 @@ switches.
+
+   By default the Secure image for the MPS2+ AN521 (CPU0) is built
+   using TF-M.
+
+
+
```
nit: Too much space

microbuilder commented on boards/arm/mps2_an521/doc/index.rst:
```diff
@@ -336,6 +423,23 @@ secure firmware:
+Building standalone applications on MPS2+ AN521 CPU1
+====================================================
+
+Applications may be build for the second Cortex-M33
```
```suggestion
Applications may be built for the second Cortex-M33
```

microbuilder commented on boards/arm/mps2_an521/CMakeLists.txt:
```diff
@@ -7,3 +7,24 @@
+# Building a firmware image for CPU1: this requires a binary
+# for CPU0, which will boot the device and wake up CPU1.
+# However, if building with OPENAMP, there is no need to build
+# any binary for CPU0, as this is build by the dual core sample.
```
```suggestion
# any binary for CPU0, as this is built by the dual core sample.
```

microbuilder commented:
Others than some minor documentation cleanup, looks good to me, and nice to finally be able to emulate FP instructions in CI. Thanks for putting this together.

ioannisg commented:
> Others than some minor documentation cleanup, looks good to me, and nice to finally be able to emulate FP instructions in CI. Thanks for putting this together.

Thanks @microbuilder , comments addressed :)

ioannisg committed:
trusted-firmware-m: rename AN521 target to BOARD_MPS2_AN521_CPU0_NS

Since we now have separate Kconfig symbols for the different
MPS2 AN521 board variants, we need to update the BOARD switch
for the TF-M target on MPS2 AN521.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

microbuilder reviewed (approved):
Tested with `scripts/twister -p mps2_an521_remote -T tests/kernel`

MartinSchoenstedt commented:
In the documentation changes you made you stated, that only CPU0 has the security extension enabled. However, for the AN521 actually both CPUs include TrustZone. I cannot find the documentation for this anymore but I have implemented a dual core system on the AN521 where both CPUs use the secure and non-secure hardware, configure their SAUs etc.

koffes commented:
@ioannisg : Great addition! I tried using this, but we have a dependency on `FP_HARDABI`. Is this supported yet?
